{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differential Expression Analysis\n",
    "This script aims to conduct differential expression analysis using Wilcoxon Rank Sum and perhaps DESeq2. The experimental groups are as follows:\n",
    " - Noninfected - PTEN WT vs PTEN KO (How does PTEN loss change overall gene expression?)\n",
    " - Infected - PTEN WT vs. PTEN KO (How does the knockout respond differently to the WT when infected?)\n",
    " - Knockouts - Non-infected vs. Infected (How does the knockout respond with and without infection?)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats, cluster\n",
    "import os\n",
    "\n",
    "#umap\n",
    "from sklearn.datasets import load_digits\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "#rna_seq analysis packages\n",
    "import pickle as pkl\n",
    "import pydeseq2 as deseq2\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.default_inference import DefaultInference\n",
    "from pydeseq2.ds import DeseqStats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Path configuration\n",
    "wd = '/Users/nguyenpu/Documents/GitHub/'\n",
    "os.chdir(wd)\n",
    "\n",
    "datadir = \"serezani_lab/bulk_rna_seq/pten_ko/data/\"\n",
    "outputdir = \"serezani_lab/bulk_rna_seq/pten_ko/outputdir\"\n",
    "\n",
    "if os.path.isdir(outputdir)==False:\n",
    "    os.mkdir(outputdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load datasets\n",
    "uninfected_df = pd.read_csv(os.path.join(datadir, 'CTLUninfected_vs_KOUninfected.genes.counts.csv'))\n",
    "infected_df = pd.read_csv(os.path.join(datadir, 'CTLInfected_vs_KOInfected.genes.counts.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load metadata\n",
    "uninfected_md = pd.read_csv(os.path.join(datadir,'uninfected_metadata.csv'), index_col = 0)\n",
    "infected_md = pd.read_csv(os.path.join(datadir,'infected_metadata.csv'), index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Name EnsemblID Column\n",
    "uninfected_df = pd.DataFrame.rename(uninfected_df, columns = {'Unnamed: 0':'Ensembl_ID'})\n",
    "infected_df = pd.DataFrame.rename(infected_df, columns = {'Unnamed: 0':'Ensembl_ID'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/k6/nqwv8_kx0zz_f6624qm3trb00000gp/T/ipykernel_53617/19316353.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ensembl2Gene_ID[['Ensembl_ID', 'delete']] = gene_info['Ensembl_ID'].str.split('.', n=1, expand=True).copy()\n",
      "/var/folders/k6/nqwv8_kx0zz_f6624qm3trb00000gp/T/ipykernel_53617/19316353.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ensembl2Gene_ID[['Ensembl_ID', 'delete']] = gene_info['Ensembl_ID'].str.split('.', n=1, expand=True).copy()\n"
     ]
    }
   ],
   "source": [
    "#Add GeneID column mapped to EnsemblID\n",
    "gene_info = pd.read_csv(os.path.join(datadir, 'gene_info'))\n",
    "ensembl2Gene_ID = gene_info[['Ensembl_ID', 'Gene_ID', 'Gene_Type']]\n",
    "\n",
    "## Only run the below lines once\n",
    "ensembl2Gene_ID[['Ensembl_ID', 'delete']] = gene_info['Ensembl_ID'].str.split('.', n=1, expand=True).copy()\n",
    "ensembl2Gene_ID = ensembl2Gene_ID.drop('delete', axis = 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl_ID</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Gene_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000094569</td>\n",
       "      <td>Trbd2</td>\n",
       "      <td>TR_D_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000095668</td>\n",
       "      <td>Trbd1</td>\n",
       "      <td>TR_D_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000096749</td>\n",
       "      <td>Trdd1</td>\n",
       "      <td>TR_D_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000096176</td>\n",
       "      <td>Trdd2</td>\n",
       "      <td>TR_D_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000094028</td>\n",
       "      <td>Ighd4-1</td>\n",
       "      <td>IG_D_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48951</th>\n",
       "      <td>ENSMUSG00000084750</td>\n",
       "      <td>Gm22453</td>\n",
       "      <td>misc_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48952</th>\n",
       "      <td>ENSMUSG00000084711</td>\n",
       "      <td>Gm24629</td>\n",
       "      <td>misc_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48953</th>\n",
       "      <td>ENSMUSG00000088014</td>\n",
       "      <td>Gm23743</td>\n",
       "      <td>misc_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48954</th>\n",
       "      <td>ENSMUSG00000099219</td>\n",
       "      <td>Gm27680</td>\n",
       "      <td>misc_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48955</th>\n",
       "      <td>ENSMUSG00000092903</td>\n",
       "      <td>Gm27896</td>\n",
       "      <td>misc_RNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48956 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Ensembl_ID  Gene_ID  Gene_Type\n",
       "0      ENSMUSG00000094569    Trbd2  TR_D_gene\n",
       "1      ENSMUSG00000095668    Trbd1  TR_D_gene\n",
       "2      ENSMUSG00000096749    Trdd1  TR_D_gene\n",
       "3      ENSMUSG00000096176    Trdd2  TR_D_gene\n",
       "4      ENSMUSG00000094028  Ighd4-1  IG_D_gene\n",
       "...                   ...      ...        ...\n",
       "48951  ENSMUSG00000084750  Gm22453   misc_RNA\n",
       "48952  ENSMUSG00000084711  Gm24629   misc_RNA\n",
       "48953  ENSMUSG00000088014  Gm23743   misc_RNA\n",
       "48954  ENSMUSG00000099219  Gm27680   misc_RNA\n",
       "48955  ENSMUSG00000092903  Gm27896   misc_RNA\n",
       "\n",
       "[48956 rows x 3 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensembl2Gene_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "ensembl2Gene_ID.to_csv(os.path.join(datadir, 'ensembl2Gene_ID.csv'), index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Merge gene name and ensembl ID\n",
    "# uninfected_df = ensembl2Gene_ID.merge(uninfected_df, on = ['Ensembl_ID'])\n",
    "# infected_df = ensembl2Gene_ID.merge(infected_df,on = ['Ensembl_ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Drop the ensembl_ID column\n",
    "# uninfected_df.drop(['Ensembl_ID'], axis = 1, inplace = True)\n",
    "# infected_df.drop(['Ensembl_ID'], axis = 1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make the Ensembl_ID column the index\n",
    "uninfected_df.set_index('Ensembl_ID', inplace = True)\n",
    "infected_df.set_index('Ensembl_ID', inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Transpose for DESeq2 analysis purposes\n",
    "uninfected = uninfected_df.transpose()\n",
    "infected = infected_df.transpose()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DeSeq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make function to perform DEseq2 analysis, so I don't need to type things twice\n",
    "def deseq_analysis(count_df, metadata, outputdir, file_name, ensembl2Gene_ID):\n",
    "    '''This function will perform analysis using DESeq2 given a count dataframe and metadata''' \n",
    "    #QC\n",
    "    genes_to_keep = count_df.columns[count_df.sum(axis=0) >= 10]\n",
    "    count_df = count_df[genes_to_keep]\n",
    "    metadata['condition'] = metadata['condition'].astype(str)\n",
    "    if 'condition' not in metadata.columns:\n",
    "        raise KeyError(\"The 'condition' column is missing from metadata!\")\n",
    "\n",
    "\n",
    "    #Create a DeseqDataSet object from the count and metadata data that were just loaded\n",
    "    inference = DefaultInference(n_cpus=6)\n",
    "    dds = DeseqDataSet(\n",
    "        counts=count_df,\n",
    "        metadata=metadata,\n",
    "        # ref_level = ['CTL', 'KO'],\n",
    "        design_factors=\"condition\",\n",
    "        refit_cooks=True,\n",
    "        inference=inference,\n",
    "    )\n",
    "\n",
    "    ##Run DESeq2 Algorithm\n",
    "    dds.deseq2()\n",
    "\n",
    "    #Statistics\n",
    "    stats_res = DeseqStats(dds, contrast=['condition','KO', 'CTL'], inference=inference) #contrast sets it so that it's WT vs. bft\n",
    "    stats_res.summary()\n",
    "\n",
    "    stats_res.lfc_shrink(coeff=None)\n",
    "    deseq2_results = stats_res.results_df\n",
    "    deseq2_results_no_NaN = deseq2_results.dropna(axis=0) #drop NA, there seems to be about 5000 genes that were dropped\n",
    "\n",
    "    # Calculate the -log10(padj) and add to the data frame\n",
    "    deseq2_results_no_NaN['-log10(padj)']=-np.log10(deseq2_results_no_NaN['padj'])\n",
    "    deseq2_results_no_NaN['-log10(p-value)']=-np.log10(deseq2_results_no_NaN['pvalue'])\n",
    "\n",
    "    #identify the genes in the dataframe that are upregulated, downregulated, or not at all\n",
    "    deseq2_results_no_NaN['DEGenes']='NO' #make a new column for the DEgenes\n",
    "\n",
    "    deseq2_results_no_NaN['DEGenes'].loc[(deseq2_results_no_NaN['log2FoldChange']>1) & (deseq2_results_no_NaN['padj']<0.05)] = 'UP'\n",
    "    deseq2_results_no_NaN['DEGenes'].loc[(deseq2_results_no_NaN['log2FoldChange']<-1) & (deseq2_results_no_NaN['padj']<0.05)] = 'DOWN'\n",
    "\n",
    "    #make and export a list of Ensembl_IDs that are upregulated and downregulated; doesn't contain additional gene annotation\n",
    "    upreg = deseq2_results_no_NaN.loc[(deseq2_results_no_NaN['DEGenes']=='UP')]\n",
    "    downreg = deseq2_results_no_NaN.loc[(deseq2_results_no_NaN['DEGenes']=='DOWN')]\n",
    "\n",
    "    deseq2_results_no_NaN.to_csv(os.path.join(outputdir,file_name))\n",
    "    return deseq2_results_no_NaN\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pydeseq2/utils.py\", line 14, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 26, in <module>\n",
      "    from pandas.compat import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/compat/__init__.py\", line 27, in <module>\n",
      "    from pandas.compat.pyarrow import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/compat/pyarrow.py\", line 8, in <module>\n",
      "    import pyarrow as pa\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pydeseq2/utils.py\", line 14, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 26, in <module>\n",
      "    from pandas.compat import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/compat/__init__.py\", line 27, in <module>\n",
      "    from pandas.compat.pyarrow import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/compat/pyarrow.py\", line 8, in <module>\n",
      "    import pyarrow as pa\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pydeseq2/utils.py\", line 14, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 26, in <module>\n",
      "    from pandas.compat import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/compat/__init__.py\", line 27, in <module>\n",
      "    from pandas.compat.pyarrow import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/compat/pyarrow.py\", line 8, in <module>\n",
      "    import pyarrow as pa\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pydeseq2/utils.py\", line 14, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 26, in <module>\n",
      "    from pandas.compat import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/compat/__init__.py\", line 27, in <module>\n",
      "    from pandas.compat.pyarrow import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/compat/pyarrow.py\", line 8, in <module>\n",
      "    import pyarrow as pa\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "Traceback (most recent call last):\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pydeseq2/utils.py\", line 14, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 26, in <module>\n",
      "    from pandas.compat import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/compat/__init__.py\", line 27, in <module>\n",
      "    from pandas.compat.pyarrow import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/compat/pyarrow.py\", line 8, in <module>\n",
      "    import pyarrow as pa\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "Traceback (most recent call last):\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pydeseq2/utils.py\", line 14, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 26, in <module>\n",
      "    from pandas.compat import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/compat/__init__.py\", line 27, in <module>\n",
      "    from pandas.compat.pyarrow import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/compat/pyarrow.py\", line 8, in <module>\n",
      "    import pyarrow as pa\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "    raise ImportError(msg)Traceback (most recent call last):\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "\n",
      "    raise ImportError(msg)    ImportErrorraise ImportError(msg)\n",
      "\n",
      ": \n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      "ImportErrorImportError: : \n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      "\n",
      "    raise ImportError(msg)\n",
      "ImportError: \n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      "    raise ImportError(msg)\n",
      "ImportError:     raise ImportError(msg)\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      "ImportError: \n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pydeseq2/utils.py\", line 14, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 49, in <module>\n",
      "    from pandas.core.api import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/core/api.py\", line 9, in <module>\n",
      "    from pandas.core.dtypes.dtypes import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/core/dtypes/dtypes.py\", line 24, in <module>\n",
      "    from pandas._libs import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pydeseq2/utils.py\", line 14, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 49, in <module>\n",
      "    from pandas.core.api import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/core/api.py\", line 9, in <module>\n",
      "    from pandas.core.dtypes.dtypes import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/core/dtypes/dtypes.py\", line 24, in <module>\n",
      "    from pandas._libs import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pydeseq2/utils.py\", line 14, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 49, in <module>\n",
      "    from pandas.core.api import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/core/api.py\", line 9, in <module>\n",
      "    from pandas.core.dtypes.dtypes import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/core/dtypes/dtypes.py\", line 24, in <module>\n",
      "    from pandas._libs import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pydeseq2/utils.py\", line 14, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 49, in <module>\n",
      "    from pandas.core.api import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/core/api.py\", line 9, in <module>\n",
      "    from pandas.core.dtypes.dtypes import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/core/dtypes/dtypes.py\", line 24, in <module>\n",
      "    from pandas._libs import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "Traceback (most recent call last):\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pydeseq2/utils.py\", line 14, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 49, in <module>\n",
      "    from pandas.core.api import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/core/api.py\", line 9, in <module>\n",
      "    from pandas.core.dtypes.dtypes import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/core/dtypes/dtypes.py\", line 24, in <module>\n",
      "    from pandas._libs import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/opt/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pydeseq2/utils.py\", line 14, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 49, in <module>\n",
      "    from pandas.core.api import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/core/api.py\", line 9, in <module>\n",
      "    from pandas.core.dtypes.dtypes import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pandas/core/dtypes/dtypes.py\", line 24, in <module>\n",
      "    from pandas._libs import (\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "    raise ImportError(msg)\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "    raise ImportError(msg)\n",
      "    raise ImportError(msg)\n",
      "        raise ImportError(msg)ImportError\n",
      ": \n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      "ImportErrorImportError: \n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      ": raise ImportError(msg)\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      "\n",
      "    raise ImportError(msg)\n",
      "ImportError: \n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      "ImportError: \n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      "ImportError: \n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.2 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.61 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.20 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.96 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.03 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 0.45 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: condition KO vs CTL\n",
      "                       baseMean  log2FoldChange     lfcSE      stat  \\\n",
      "Ensembl_ID                                                            \n",
      "ENSMUSG00000000001  9027.509719       -0.337508  0.096515 -3.496954   \n",
      "ENSMUSG00000000028  1346.083231       -0.197182  0.118203 -1.668162   \n",
      "ENSMUSG00000000056   923.275989        0.154391  0.132648  1.163911   \n",
      "ENSMUSG00000000058  3078.829487        0.062695  0.122168  0.513186   \n",
      "ENSMUSG00000000078  8236.341593       -0.577460  0.101141 -5.709427   \n",
      "...                         ...             ...       ...       ...   \n",
      "ENSMUSG00000118591    10.568534        2.904650  0.804464  3.610663   \n",
      "ENSMUSG00000118607   202.616075        0.894560  0.189092  4.730818   \n",
      "ENSMUSG00000118618     1.542217        3.970720  2.552247  1.555774   \n",
      "ENSMUSG00000118619     4.196095       -0.031968  1.250840 -0.025557   \n",
      "ENSMUSG00000118623     2.101578       -3.525727  2.016014 -1.748860   \n",
      "\n",
      "                          pvalue          padj  \n",
      "Ensembl_ID                                      \n",
      "ENSMUSG00000000001  4.706021e-04  1.602477e-03  \n",
      "ENSMUSG00000000028  9.528367e-02  1.723917e-01  \n",
      "ENSMUSG00000000056  2.444600e-01  3.677458e-01  \n",
      "ENSMUSG00000000058  6.078209e-01  7.233372e-01  \n",
      "ENSMUSG00000000078  1.133572e-08  8.100136e-08  \n",
      "...                          ...           ...  \n",
      "ENSMUSG00000118591  3.054150e-04  1.079898e-03  \n",
      "ENSMUSG00000118607  2.236171e-06  1.163798e-05  \n",
      "ENSMUSG00000118618  1.197618e-01           NaN  \n",
      "ENSMUSG00000118619  9.796107e-01  9.873786e-01  \n",
      "ENSMUSG00000118623  8.031518e-02           NaN  \n",
      "\n",
      "[18491 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 1.93 seconds.\n",
      "\n",
      "/opt/anaconda3/lib/python3.12/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n",
      "/var/folders/k6/nqwv8_kx0zz_f6624qm3trb00000gp/T/ipykernel_53617/2962324808.py:35: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  deseq2_results_no_NaN['-log10(padj)']=-np.log10(deseq2_results_no_NaN['padj'])\n",
      "/opt/anaconda3/lib/python3.12/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n",
      "/var/folders/k6/nqwv8_kx0zz_f6624qm3trb00000gp/T/ipykernel_53617/2962324808.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  deseq2_results_no_NaN['-log10(p-value)']=-np.log10(deseq2_results_no_NaN['pvalue'])\n",
      "/var/folders/k6/nqwv8_kx0zz_f6624qm3trb00000gp/T/ipykernel_53617/2962324808.py:39: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  deseq2_results_no_NaN['DEGenes']='NO' #make a new column for the DEgenes\n",
      "/var/folders/k6/nqwv8_kx0zz_f6624qm3trb00000gp/T/ipykernel_53617/2962324808.py:41: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  deseq2_results_no_NaN['DEGenes'].loc[(deseq2_results_no_NaN['log2FoldChange']>1) & (deseq2_results_no_NaN['padj']<0.05)] = 'UP'\n",
      "/var/folders/k6/nqwv8_kx0zz_f6624qm3trb00000gp/T/ipykernel_53617/2962324808.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  deseq2_results_no_NaN['DEGenes'].loc[(deseq2_results_no_NaN['log2FoldChange']<-1) & (deseq2_results_no_NaN['padj']<0.05)] = 'DOWN'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: condition KO vs CTL\n",
      "                       baseMean  log2FoldChange     lfcSE      stat  \\\n",
      "Ensembl_ID                                                            \n",
      "ENSMUSG00000000001  9027.509719       -0.337244  0.096110 -3.496954   \n",
      "ENSMUSG00000000028  1346.083231       -0.190714  0.117412 -1.668162   \n",
      "ENSMUSG00000000056   923.275989        0.145305  0.131513  1.163911   \n",
      "ENSMUSG00000000058  3078.829487        0.057964  0.121195  0.513186   \n",
      "ENSMUSG00000000078  8236.341593       -0.573360  0.100945 -5.709427   \n",
      "...                         ...             ...       ...       ...   \n",
      "ENSMUSG00000118591    10.568534        2.460562  0.799310  3.610663   \n",
      "ENSMUSG00000118607   202.616075        0.857482  0.188980  4.730818   \n",
      "ENSMUSG00000118618     1.542217        0.413443  1.739687  1.555774   \n",
      "ENSMUSG00000118619     4.196095       -0.006363  0.758379 -0.025557   \n",
      "ENSMUSG00000118623     2.101578       -0.457568  0.939469 -1.748860   \n",
      "\n",
      "                          pvalue          padj  \n",
      "Ensembl_ID                                      \n",
      "ENSMUSG00000000001  4.706021e-04  1.602477e-03  \n",
      "ENSMUSG00000000028  9.528367e-02  1.723917e-01  \n",
      "ENSMUSG00000000056  2.444600e-01  3.677458e-01  \n",
      "ENSMUSG00000000058  6.078209e-01  7.233372e-01  \n",
      "ENSMUSG00000000078  1.133572e-08  8.100136e-08  \n",
      "...                          ...           ...  \n",
      "ENSMUSG00000118591  3.054150e-04  1.079898e-03  \n",
      "ENSMUSG00000118607  2.236171e-06  1.163798e-05  \n",
      "ENSMUSG00000118618  1.197618e-01           NaN  \n",
      "ENSMUSG00000118619  9.796107e-01  9.873786e-01  \n",
      "ENSMUSG00000118623  8.031518e-02           NaN  \n",
      "\n",
      "[18491 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.86 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.20 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 2.05 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.10 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 0.52 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: condition KO vs CTL\n",
      "                       baseMean  log2FoldChange     lfcSE      stat  \\\n",
      "Ensembl_ID                                                            \n",
      "ENSMUSG00000000001  8271.732962       -0.179003  0.066674 -2.684733   \n",
      "ENSMUSG00000000028   982.445119       -0.271353  0.107660 -2.520472   \n",
      "ENSMUSG00000000037     1.987273        1.590849  2.002825  0.794302   \n",
      "ENSMUSG00000000056  1007.383977        0.685686  0.099740  6.874736   \n",
      "ENSMUSG00000000058  2663.673954        0.390209  0.084888  4.596759   \n",
      "...                         ...             ...       ...       ...   \n",
      "ENSMUSG00000118590    72.775124        1.388921  0.282500  4.916535   \n",
      "ENSMUSG00000118591    14.299576        3.213915  0.748050  4.296390   \n",
      "ENSMUSG00000118607   301.326762        1.455751  0.185192  7.860785   \n",
      "ENSMUSG00000118619     4.232791        0.352490  1.033357  0.341111   \n",
      "ENSMUSG00000118640     2.485392        0.223911  1.624520  0.137832   \n",
      "\n",
      "                          pvalue          padj  \n",
      "Ensembl_ID                                      \n",
      "ENSMUSG00000000001  7.258770e-03  1.468690e-02  \n",
      "ENSMUSG00000000028  1.171975e-02  2.286898e-02  \n",
      "ENSMUSG00000000037  4.270194e-01  5.331609e-01  \n",
      "ENSMUSG00000000056  6.210491e-12  3.000129e-11  \n",
      "ENSMUSG00000000058  4.291128e-06  1.285984e-05  \n",
      "...                          ...           ...  \n",
      "ENSMUSG00000118590  8.808961e-07  2.816307e-06  \n",
      "ENSMUSG00000118591  1.736018e-05  4.885988e-05  \n",
      "ENSMUSG00000118607  3.817331e-15  2.340483e-14  \n",
      "ENSMUSG00000118619  7.330199e-01  8.054467e-01  \n",
      "ENSMUSG00000118640  8.903733e-01  9.231795e-01  \n",
      "\n",
      "[18395 rows x 6 columns]\n",
      "Shrunk log2 fold change & Wald test p-value: condition KO vs CTL\n",
      "                       baseMean  log2FoldChange     lfcSE      stat  \\\n",
      "Ensembl_ID                                                            \n",
      "ENSMUSG00000000001  8271.732962       -0.171397  0.066526 -2.684733   \n",
      "ENSMUSG00000000028   982.445119       -0.265849  0.107188 -2.520472   \n",
      "ENSMUSG00000000037     1.987273        0.194692  1.162100  0.794302   \n",
      "ENSMUSG00000000056  1007.383977        0.677375  0.099607  6.874736   \n",
      "ENSMUSG00000000058  2663.673954        0.386504  0.084683  4.596759   \n",
      "...                         ...             ...       ...       ...   \n",
      "ENSMUSG00000118590    72.775124        1.315805  0.282911  4.916535   \n",
      "ENSMUSG00000118591    14.299576        2.880454  0.744195  4.296390   \n",
      "ENSMUSG00000118607   301.326762        1.425024  0.185367  7.860785   \n",
      "ENSMUSG00000118619     4.232791        0.123231  0.755963  0.341111   \n",
      "ENSMUSG00000118640     2.485392        0.040121  0.898797  0.137832   \n",
      "\n",
      "                          pvalue          padj  \n",
      "Ensembl_ID                                      \n",
      "ENSMUSG00000000001  7.258770e-03  1.468690e-02  \n",
      "ENSMUSG00000000028  1.171975e-02  2.286898e-02  \n",
      "ENSMUSG00000000037  4.270194e-01  5.331609e-01  \n",
      "ENSMUSG00000000056  6.210491e-12  3.000129e-11  \n",
      "ENSMUSG00000000058  4.291128e-06  1.285984e-05  \n",
      "...                          ...           ...  \n",
      "ENSMUSG00000118590  8.808961e-07  2.816307e-06  \n",
      "ENSMUSG00000118591  1.736018e-05  4.885988e-05  \n",
      "ENSMUSG00000118607  3.817331e-15  2.340483e-14  \n",
      "ENSMUSG00000118619  7.330199e-01  8.054467e-01  \n",
      "ENSMUSG00000118640  8.903733e-01  9.231795e-01  \n",
      "\n",
      "[18395 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 2.20 seconds.\n",
      "\n",
      "/opt/anaconda3/lib/python3.12/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n",
      "/var/folders/k6/nqwv8_kx0zz_f6624qm3trb00000gp/T/ipykernel_53617/2962324808.py:35: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  deseq2_results_no_NaN['-log10(padj)']=-np.log10(deseq2_results_no_NaN['padj'])\n",
      "/opt/anaconda3/lib/python3.12/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n",
      "/var/folders/k6/nqwv8_kx0zz_f6624qm3trb00000gp/T/ipykernel_53617/2962324808.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  deseq2_results_no_NaN['-log10(p-value)']=-np.log10(deseq2_results_no_NaN['pvalue'])\n",
      "/var/folders/k6/nqwv8_kx0zz_f6624qm3trb00000gp/T/ipykernel_53617/2962324808.py:39: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  deseq2_results_no_NaN['DEGenes']='NO' #make a new column for the DEgenes\n",
      "/var/folders/k6/nqwv8_kx0zz_f6624qm3trb00000gp/T/ipykernel_53617/2962324808.py:41: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  deseq2_results_no_NaN['DEGenes'].loc[(deseq2_results_no_NaN['log2FoldChange']>1) & (deseq2_results_no_NaN['padj']<0.05)] = 'UP'\n",
      "/var/folders/k6/nqwv8_kx0zz_f6624qm3trb00000gp/T/ipykernel_53617/2962324808.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  deseq2_results_no_NaN['DEGenes'].loc[(deseq2_results_no_NaN['log2FoldChange']<-1) & (deseq2_results_no_NaN['padj']<0.05)] = 'DOWN'\n"
     ]
    }
   ],
   "source": [
    "#Run DESeq2 in terms of KO vs. CTL\n",
    "\n",
    "uninfected_deseq2 = deseq_analysis(uninfected, uninfected_md, outputdir, 'DESeq2_CTLUninfected_vs_KOUninfected.csv', ensembl2Gene_ID)\n",
    "infected_deseq2 = deseq_analysis(infected, infected_md, outputdir, 'DESeq2_CTLInfected_vs_KOInfected.csv', ensembl2Gene_ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "ensembl2Gene_ID = pd.read_csv(os.path.join(datadir, 'ensembl2Gene_ID.csv'))\n",
    "\n",
    "#Merge gene name and ensembl ID\n",
    "uninfected_deseq2_final = ensembl2Gene_ID.merge(uninfected_deseq2, on = 'Ensembl_ID')\n",
    "uninfected_deseq2_final.to_csv(os.path.join(outputdir, 'DESeq2_CTLUninfected_vs_KOUninfected.csv'), index = False)\n",
    "\n",
    "infected_deseq2_final = ensembl2Gene_ID.merge(infected_deseq2, on = 'Ensembl_ID')\n",
    "infected_deseq2_final.to_csv(os.path.join(outputdir, 'DESeq2_CTLInfected_vs_KOInfected.csv'), index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make lists for upregulated and downregulated genes only\n",
    "upreg_uninfected = uninfected_deseq2_final.loc[(uninfected_deseq2_final['DEGenes']=='UP')]\n",
    "downreg_uninfected = uninfected_deseq2_final.loc[(uninfected_deseq2_final['DEGenes']=='DOWN')]\n",
    "\n",
    "upreg_infected = infected_deseq2_final.loc[(infected_deseq2_final['DEGenes']=='UP')]\n",
    "downreg_infected = infected_deseq2_final.loc[(infected_deseq2_final['DEGenes']=='DOWN')]\n",
    "\n",
    "#Sort the dataframes by abs(log2FoldChange), greatest to least\n",
    "upreg_uninfected = upreg_uninfected.sort_values(by = 'log2FoldChange', ascending = False)\n",
    "downreg_uninfected = downreg_uninfected.sort_values(by = 'log2FoldChange', ascending = True)\n",
    "\n",
    "upreg_infected = upreg_infected.sort_values(by = 'log2FoldChange', ascending = False)\n",
    "downreg_infected = downreg_infected.sort_values(by = 'log2FoldChange', ascending = True)\n",
    "\n",
    "#Save the upregulated and downregulated genes\n",
    "upreg_uninfected.to_csv(os.path.join(outputdir, 'upreg_uninfected.csv'), index = False)\n",
    "downreg_uninfected.to_csv(os.path.join(outputdir, 'downreg_uninfected.csv'), index = False)\n",
    "\n",
    "upreg_infected.to_csv(os.path.join(outputdir, 'upreg_infected.csv'), index = False)\n",
    "downreg_infected.to_csv(os.path.join(outputdir, 'downreg_infected.csv'), index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
